open_spiel/games/phantom_go.cc

```diff
@@ -0,0 +1,301 @@
+     // score is computed as usual (i.e. number of stones + komi).
+     // It's advised to only use shorter games to compute win-rates.
+     // When not provided, it defaults to DefaultMaxGameLength(board_size)
+     {"max_game_length",
+      GameParameter(GameParameter::Type::kInt, /*is_mandatory=*/false)}},
+};
+
+std::shared_ptr<const Game> Factory(const GameParameters& params) {
+  return std::shared_ptr<const Game>(new PhantomGoGame(params));
+}
+
+REGISTER_SPIEL_GAME(kGameType, Factory);
+
+std::vector<VirtualPoint> HandicapStones(int num_handicap) {
+  if (num_handicap < 2 || num_handicap > 9) return {};
+
+  static std::array<VirtualPoint, 9> placement = {
+      {MakePoint("d4"), MakePoint("q16"), MakePoint("d16"), MakePoint("q4"),
+       MakePoint("d10"), MakePoint("q10"), MakePoint("k4"), MakePoint("k16"),
+       MakePoint("k10")}};
+  static VirtualPoint center = MakePoint("k10");
+
+  std::vector<VirtualPoint> points;
+  points.reserve(num_handicap);
+  for (int i = 0; i < num_handicap; ++i) {
+    points.push_back(placement[i]);
+  }
+
+  if (num_handicap >= 5 && num_handicap % 2 == 1) {
+    points[num_handicap - 1] = center;
+  }
+
+  return points;
+}
+
+}  // namespace
+
+PhantomGoState::PhantomGoState(std::shared_ptr<const Game> game, int board_size, float komi,
+                 int handicap)
+    //help 
+    : State(std::move(game)),
+      board_(board_size),
+      komi_(komi),
+      handicap_(handicap),
+      max_game_length_(game_->MaxGameLength()),
+      to_play_(GoColor::kBlack) {
+  ResetBoard();
+  
+}
+
+std::unique_ptr<State> PhantomGoState::ResampleFromInfostate(
+    int player_id, std::function<double()> rng) const {
+    int boardSize = board_.board_size();
+
+    std::shared_ptr<const Game> newGame = LoadGame("phantom_go");
+    std::unique_ptr<PhantomGoState> newState = std::make_unique<PhantomGoState>(PhantomGoState(newGame, boardSize, komi_, handicap_));
+
+    std::array<GoColor, kMaxBoardSize* kMaxBoardSize> infoState = board_.getObservationByID(player_id);
+    std::array<int, 2> stoneCount = board_.getStoneCount();
+    int historyLength = history_.size();
+    int enemyStonesPlaced = 0;
+
+    // Replicate all visible stones
+    for (int i = 0; i < boardSize * boardSize; i++)
+    {
+        if (infoState[i] != GoColor::kEmpty)
+        {
+            newState->board_.PlayMove(VirtualPointFromBoardPoint(i, boardSize), infoState[i]);
+            if ((uint8_t)OppColor(infoState[i]) == player_id)
+            {
+                enemyStonesPlaced++;
+                //newState->board_.addEnemyStoneIntoObservation(i, player_id);
+            }
+        }
+    }
+
+
+
+    return newState;
+}
+
+std::string PhantomGoState::InformationStateString(int player) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+  return HistoryString();
+}
+
+std::string PhantomGoState::ObservationString(int player) const {
```

Comment:
Similar as for InformationStateString, this is not a correct implementation.

Reply:
I don't think this is correct yet either: only board is displayed as observation, but there are other messages from the referee (such as invalid move).